src/dune_rules/dir_contents.ml

```diff
@@ -469,10 +469,14 @@ include Load
 
 let modules_of_lib sctx lib =
   let dir = Lib_info.src_dir (Lib.info lib) in
-  match Path.as_in_build_dir dir with
-  | None -> Memo.return None
-  | Some dir ->
-    let* t = get sctx ~dir in
-    let+ ml_sources = ocaml t in
-    let name = Lib.name lib in
-    Some (Ml_sources.modules ml_sources ~for_:(Library name))
+  let info = Lib.info lib in
+  match Lib_info.modules info with
+  | External modules -> Memo.return (Some modules)
+  | Local -> (
+    match Path.as_in_build_dir dir with
```

Comment:
should I be using `Path.as_in_build_dir_exn` here?

Reply:
Good question. The answer is no because it will fail for findlib libraries. We need to improve the types per my comment https://github.com/ocaml/dune/pull/6605#discussion_r1036059790 or Ali's suggestion https://github.com/ocaml/dune/pull/6605#issuecomment-1332292887